airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumStateDecoratingIterator.java

```diff
@@ -0,0 +1,179 @@
+   */
+  public DebeziumStateDecoratingIterator(final Iterator<ChangeEvent<String, String>> changeEventIterator,
+                                         final CdcStateHandler cdcStateHandler,
+                                         final CdcMetadataInjector cdcMetadataInjector,
+                                         final Instant emittedAt,
+                                         final AirbyteFileOffsetBackingStore offsetManager,
+                                         final boolean trackSchemaHistory,
+                                         final AirbyteSchemaHistoryStorage schemaHistoryManager,
+                                         final Integer checkpointSeconds,
+                                         final Integer checkpointRecords) {
+    this.changeEventIterator = changeEventIterator;
+    this.cdcStateHandler = cdcStateHandler;
+    this.cdcMetadataInjector = cdcMetadataInjector;
+    this.emittedAt = emittedAt;
+    this.offsetManager = offsetManager;
+    this.trackSchemaHistory = trackSchemaHistory;
+    this.schemaHistoryManager = schemaHistoryManager;
+
+    this.syncCheckpointDuration = Duration.ofSeconds(checkpointSeconds);
+    this.syncCheckpointRecords = checkpointRecords;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    resetCheckpointValues();
+  }
+
+  /**
+   * Computes the next record retrieved from Source stream. Emits state messages as checkpoints based
+   * on number of records or time lapsed.
+   *
+   * <p>
+   * If this method throws an exception, it will propagate outward to the {@code hasNext} or
+   * {@code next} invocation that invoked this method. Any further attempts to use the iterator will
+   * result in an {@link IllegalStateException}.
+   * </p>
+   *
+   * @return {@link AirbyteStateMessage} containing CDC data or state checkpoint message.
+   */
+  @Override
+  protected AirbyteMessage computeNext() {
+    if (isSyncFinished) {
+      return endOfData();
+    }
+
+    if (sendCheckpointMessage) {
+      AirbyteMessage stateMessage = createStateMessage(checkpointOffset);
+      previousCheckpointOffset.clear();
+      previousCheckpointOffset.putAll(checkpointOffset);
+      resetCheckpointValues();
+      return stateMessage;
+    }
+
+    if (changeEventIterator.hasNext()) {
+      try {
+        final ChangeEvent<String, String> event = changeEventIterator.next();
+        recordsLastSync++;
+
+        if (checkpointOffset.size() == 0 &&
+            (recordsLastSync >= syncCheckpointRecords ||
+                Duration.between(dateTimeLastSync, OffsetDateTime.now()).compareTo(syncCheckpointDuration) > 0)) {
+          checkpointOffset.putAll(offsetManager.read());
+          if (!previousCheckpointOffset.isEmpty() &&
+              cdcStateHandler.isSameOffset(checkpointOffset, previousCheckpointOffset)) {
+            checkpointOffset.clear();
+          }
+        }
+
+        if (checkpointOffset.size() == 1
```

Comment:
Can you explain what is this checkpointOffset

Reply:
I will add some comments on the variable and probably it's easier to rename to `checkpointOffsetToSend` so it's clearer that it's the one we can't to propagate.

Reply:
Thank you. I get it now.
Looking good